system/Traits/ConditionalTrait.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Traits;
+
+trait ConditionalTrait
+{
+    /**
+     * Only runs the query when $condition evaluates to true
+     *
+     * @param array|bool|float|int|object|resource|string|null $condition
+     *
+     * @return $this
+     */
+    public function when($condition, callable $callback, ?callable $defaultCallback = null): self
+    {
+        if ($condition) {
+            $callback($this, $condition);
+        } elseif ($defaultCallback) {
+            $defaultCallback($this);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Only runs the query when $condition evaluates to false
+     *
+     * @param array|bool|float|int|object|resource|string|null $condition
+     *
+     * @return $this
+     */
+    public function whenNot($condition, callable $callback, ?callable $defaultCallback = null): self
```

Comment:
`whenNot` as the inverse of `when` seems quite not natural for me. It's a bit off. How about other antonyms of `when`:

- except
- unless
- if not
- lest

Reply:
The method expects the condition to evaluate to `false`, which makes `except` and `unless` not make semantic sense. `ifNot` as an opposite of `when` feels wrong also. `lest` feels like the condition should evaluate to `true` to me, also. 

`whenNot` does at least match the pattern of other methods like `where`/`whereNot`. I'm open to suggestions but I'm not sure any of these fit, unless we change the whole method to suit the name.

Reply:
`except`/`unless` makes sense to me if the statement is read in a specific way. In the context of the DB builder, since I know the usage of `when` then when I read its inverse (`except`/`unless`):
```php
$builder = $this->db->table('users');
$user = model('Users')->find(1);

return $builder->except($user, fn ($query) { /* some code */ })->get();
```
In this example, I would read the code as "Except when `$user` is truthy (or a found user record), let the `$builder` execute the callable then get the DB result."

Reply:
`whenNot` is easy to understand for non native English speakers.
It is logical. The `when` `whenNot` pair is easy to remember, because we already have where/whereNot.

`except` is difficult to imagine as the opposite of `when`.